src/data_source.rs

```diff
@@ -0,0 +1,101 @@
+//! Holds implementation of odbc connection
+use super::{raw, Environment, Result, DiagRec, Error};
+use std;
+use std::marker::PhantomData;
+
+/// Represents a connection to an ODBC data source
+pub struct DataSource<'a> {
+    handle: raw::SQLHDBC,
+    // we use phantom data to tell the borrow checker that we need to keep the environment alive for
+    // the lifetime of the connection
+    env: PhantomData<&'a Environment>,
+}
+
+impl<'a> DataSource<'a> {
+    /// Connects to an ODBC data source
+    ///
+    /// # Arguments
+    /// * `env` - Environment used to allocate the data source handle.
+    /// * `dsn` - Data source name configured in the `odbc.ini` file!()
+    /// * `usr` - User identifier
+    /// * `pwd` - Authentication (usually password)
+    pub fn with_dsn_and_credentials<'b>(env: &'b mut Environment,
+                                        dsn: &str,
```

Comment:
Thanks for the feedback; The article made for an interesting read. Different from the situation displayed in the article we do not have the need to take ownership of the string, nor do we copy it. In both cases, a plain slice `&str` as well as an owned `String` we can offer reasonable ergonomics without any performance penalties. I compiled both snippets locally to be sure they work. Owned `String`:
```rust
let dsn = String::from("PostgreSQL");
let conn = DataSource::with_dsn_and_credentials(&mut environment, &dsn, "postgres", "")
            .expect("Could not connect");
```
plain `&str`
```rust
let dsn = "PostgreSQL";
let conn = DataSource::with_dsn_and_credentials(&mut environment, dsn, "postgres", "")
            .expect("Could not connect");
```
Apart from the extra `&` so coercion kicks in, owned `String`s require no additional fuss. What do you think?